replicator/active_replicator_pull.go

```diff
@@ -69,19 +162,163 @@ func (apr *ActivePullReplicator) Start() error {
 		return err
 	}
 
+	checkpoint := apr.getCheckpoint()
+	apr.checkpointRevID = checkpoint.RevID
+
 	subChangesRequest := SubChangesRequest{
 		Continuous:     apr.config.Continuous,
 		Batch:          apr.config.ChangesBatchSize,
-		Since:          apr.config.Since,
+		Since:          checkpoint.Checkpoint.LastSequence,
 		Filter:         apr.config.Filter,
 		FilterChannels: apr.config.FilterChannels,
 		DocIDs:         apr.config.DocIDs,
 		ActiveOnly:     apr.config.ActiveOnly,
 	}
 
+	if err := apr.startCheckpointer(); err != nil {
+		return err
+	}
+
 	if err := subChangesRequest.Send(apr.blipSender); err != nil {
 		return err
 	}
 
 	return nil
 }
+
+// startCheckpointer registers appropriate callback functions for checkpointing, and starts a time-based checkpointer goroutine.
+func (apr *ActivePullReplicator) startCheckpointer() error {
+	if apr.config.CheckpointMinInterval == 0 {
+		apr.config.CheckpointMinInterval = time.Second * 10
+	}
+	if apr.config.CheckpointMaxInterval == 0 {
+		apr.config.CheckpointMaxInterval = time.Second * 30
+	}
+	if apr.config.CheckpointRevCount == 0 {
+		apr.config.CheckpointRevCount = apr.config.ChangesBatchSize
+	}
+
+	// checkpointerLock guards the expectedSeqs slice, and receivedSeqs map
+	checkpointerLock := sync.Mutex{}
+
+	// An ordered list of sequence IDs we expect to receive revs for.
+	expectedSeqs := []db.SequenceID{}
+	// A map of sequence IDs we've received revs for.
+	receivedSeqs := map[db.SequenceID]struct{}{}
+
+	checkpointNow := make(chan struct{})
+
+	apr.blipSyncContext.postHandleChangesCallback = func(changesSeqs []db.SequenceID) {
+		apr.Stats.Add(ActiveReplicatorStatsKeyChangesRevsReceivedTotal, int64(len(changesSeqs)))
+
+		if len(changesSeqs) == 0 {
+			// nothing to do
+			return
+		}
+
+		select {
+		case <-apr.ctx.Done():
+			// replicator already closed, bail out of checkpointing work
+			return
+		default:
+		}
+
+		checkpointerLock.Lock()
+		expectedSeqs = append(expectedSeqs, changesSeqs...)
+		checkpointerLock.Unlock()
+	}
+
+	revCounter := uint16(0)
+	apr.blipSyncContext.postHandleRevCallback = func(remoteSeq db.SequenceID) {
+		apr.Stats.Add(ActiveReplicatorStatsKeyRevsReceivedTotal, 1)
+
+		select {
+		case <-apr.ctx.Done():
+			// replicator already closed, bail out of checkpointing work
+			return
+		default:
+		}
+
+		checkpointerLock.Lock()
+		revCounter++
+		receivedSeqs[remoteSeq] = struct{}{}
+
+		shouldCheckpoint := revCounter >= apr.config.CheckpointRevCount
+		if !shouldCheckpoint {
+			checkpointerLock.Unlock()
+			return
+		}
+
+		revCounter = uint16(0)
+		checkpointerLock.Unlock()
+
+		select {
+		case checkpointNow <- struct{}{}:
+		case <-time.After(apr.config.CheckpointMinInterval):
```

Comment:
I need to rework this throttle, because it's going to allow checkpoints to happen one after another without a time in between.